tfhe/src/integer/server_key/radix_parallel/sub.rs

```diff
@@ -402,6 +413,329 @@ impl ServerKey {
+                        2 << 1 // Generates a borrow
+                    } else {
+                        0 // Nothing
+                    }
+                };
+                self.key
+                    .generate_many_lookup_table(&[&first_block_state_fn, &shift_block_fn])
+            } else {
+                first_grouping_luts[2].clone()
+            }
+        };
+
+        let tmp = blocks
+            .par_iter()
+            .enumerate()
+            .map(|(index, block)| {
+                let grouping_index = index / grouping_size;
+                let is_in_first_grouping = grouping_index == 0;
+                let index_in_grouping = index % (grouping_size);
+                let is_last_index = index == blocks.len() - 1;
+
+                let luts = if is_last_index {
+                    &last_block_luts
+                } else if is_in_first_grouping {
+                    &first_grouping_luts[index_in_grouping]
+                } else {
+                    &other_block_state_luts[index_in_grouping]
+                };
+                self.key.apply_many_lookup_table(block, luts)
+            })
+            .collect::<Vec<_>>();
+
+        let mut shifted_blocks = Vec::with_capacity(num_blocks);
+        let mut block_states = Vec::with_capacity(num_blocks);
+        for mut blocks in tmp {
+            assert_eq!(blocks.len(), 2);
+            shifted_blocks.push(blocks.pop().unwrap());
+            block_states.push(blocks.pop().unwrap());
+        }
+
+        (shifted_blocks, block_states)
+    }
+
+    pub(crate) fn advanced_sub_assign_with_borrow_sequential(
+        &self,
+        lhs: &mut RadixCiphertext,
+        rhs: &RadixCiphertext,
+        input_borrow: Option<&BooleanBlock>,
+        compute_overflow: bool,
+    ) -> Option<BooleanBlock> {
+        assert_eq!(
+            lhs.blocks.len(),
+            rhs.blocks.len(),
+            "Left hand side must must have a number of blocks equal \
+            to the number of blocks of the right hand side: lhs {} blocks, rhs {} blocks",
+            lhs.blocks.len(),
+            rhs.blocks.len()
+        );
+
+        let modulus = self.key.message_modulus.0 as u64;
+
+        // If the block does not have a carry after the subtraction, it means it needs to
+        // borrow from the next block
+        let compute_borrow_lut = self
+            .key
+            .generate_lookup_table(|x| if x < modulus { 1 } else { 0 });
+
+        let mut borrow = input_borrow.map_or_else(|| self.key.create_trivial(0), |b| b.0.clone());
+        for (lhs_block, rhs_block) in lhs.blocks.iter_mut().zip(rhs.blocks.iter()) {
+            self.key.unchecked_sub_assign(lhs_block, rhs_block);
+            // Here unchecked_sub_assign does not give correct result, we don't want
+            // the correcting term to be used
+            // -> This is ok as the value returned by unchecked_sub is in range
+            // 1..(message_mod * 2)
```

Comment:
not sure I'm clear on why this is ok

Reply:
sub assign is (x - y + msg_mod), so with (3-0 + 4) = 7, and (0 - 3 + 4) = 1, the resulting ciphertext is in range 1..msg_mod